File: src/main/java/com/lycoon/clashapi/core/CoreUtils.java
package com.lycoon.clashapi.core;

import com.google.gson.Gson;
import com.lycoon.clashapi.core.exception.*;
import okhttp3.Response;

import java.io.IOException;
import java.util.Objects;

public class CoreUtils
{
    public final static String URL = "https://api.clashofclans.com/";
    public final static String API_VERSION = "v1";
    final static Gson gson = new Gson();

    public static Response checkResponse(Response res) throws ClashAPIException
    {
        if (res.isSuccessful())
            return res;

        switch(res.code())
        {
            case 400:
                throw new BadRequestException();
            case 403:
                throw new AuthException();
            case 404:
                throw new NotFoundException();
            case 429:
                throw new RateLimitException();
            case 503:
                throw new MaintenanceException();
            default:
                throw new UnknownException();
        }
    }

    public static <T> T deserialize(Response res, Class<T> obj) throws IOException
    {
        T deserialized = gson.fromJson(Objects.requireNonNull(res.body()).string(), obj);
        res.close();

        return deserialized;
    }

    public static String formatTag(String tag)
    {
        if (tag.startsWith("#"))
            return tag.replace("#", "%23");
        return "%23" + tag;
    }
}

File: src/main/java/com/lycoon/clashapi/core/CoreUtils.kt
package com.lycoon.clashapi.core

import com.lycoon.clashapi.core.exceptions.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException

object CoreUtils {
    const val JSON_CONTENT_TYPE = "application/json; charset=utf-8"
    const val API_URL = "https://api.clashofclans.com/"
    const val API_VERSION = "v1"

    val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        coerceInputValues = true
        prettyPrint = true
    }

    @Throws(IOException::class)
    inline fun <reified T> deserialize(res: Response): T {
        return json.decodeFromString(res.body?.string() ?: "")
    }

    @Throws(IOException::class)
    inline fun <reified T> serialize(obj: T): String {
        return json.encodeToString<T>(obj)
    }

    /*
     * There are situations where we get a list from the API, but we can't return it directly as List<T>
     * because the API returns that list inside an object, within a single attribute named 'items' and
     * Kotlinx Serialization doesn't support that.
     *
     * For the concerned routes, we unwrap the list to avoid duplicating useless single-attribute classes
     * wrapping the list.
     */
    @Serializable
    class WrapperList<T>(val items: List<T>)
    fun <T> unwrapList(obj: WrapperList<T>): List<T> { return obj.items }

    /*
     * We don't want to invalidate a request if the provided tag doesn't start with a '#', so we prepend it
     * if it's not present. The %23 is the URL encoded version of '#'.
     */
    fun formatTag(tag: String): String{
        return if (tag.startsWith("#")) tag.replace("#", "%23") else "%23$tag"
    }

    /*
     * Prepares a RequestBody from a data class
     */
    inline fun <reified T> getRequestBody(obj: T): RequestBody {
        val body = serialize(obj)
        return body.toRequestBody(JSON_CONTENT_TYPE.toMediaTypeOrNull())
    }

    @Throws(ClashAPIException::class)
    fun checkResponse(res: Response): Response {
        if (res.isSuccessful) return res
        when (res.code) {
            400 -> throw BadRequestException()
            403 -> throw AuthException()
            404 -> throw NotFoundException()
            429 -> throw RateLimitException()
            503 -> throw MaintenanceException()
            else -> throw UnknownException()
        }
    }
}